File: sdk/eventbus/eventbus.go
package eventbus

import (
	"log"
	"sync"

	"hei-gin/api"
)

// DefaultBus is the global event bus instance.
var DefaultBus api.EventBus = newBus()

type bus struct {
	mu          sync.RWMutex
	subscribers map[string][]api.EventSubscriber
	closed      chan struct{}
	wg          sync.WaitGroup
}

func newBus() *bus {
	return &bus{
		subscribers: make(map[string][]api.EventSubscriber),
		closed:      make(chan struct{}),
	}
}

func (b *bus) Publish(topic string, data any) {
	b.mu.RLock()
	subs, ok := b.subscribers[topic]
	b.mu.RUnlock()
	if !ok {
		return
	}
	event := api.Event{Topic: topic, Data: data}
	for _, sub := range subs {
		sub := sub
		b.wg.Add(1)
		go func() {
			defer b.wg.Done()
			defer func() {
				if r := recover(); r != nil {
					log.Printf("[EventBus] subscriber panic on topic %s: %v", topic, r)
				}
			}()
			select {
			case <-b.closed:
				return
			default:
				sub(event)
			}
		}()
	}
}

func (b *bus) Subscribe(topic string, sub api.EventSubscriber) func() {
	b.mu.Lock()
	b.subscribers[topic] = append(b.subscribers[topic], sub)
	b.mu.Unlock()

	return func() {
		b.mu.Lock()
		defer b.mu.Unlock()
		subs := b.subscribers[topic]
		for i, s := range subs {
			if &s == &sub {
				b.subscribers[topic] = append(subs[:i], subs[i+1:]...)
				return
			}
		}
	}
}

// Topics used across the system.
const (
	TopicUserConnected    = "user:connected"
	TopicUserDisconnected = "user:disconnected"
	TopicMessageSent      = "message:sent"
	TopicMessageRead      = "message:read"
)

File: sdk/infra/eventbus/eventbus.py
from __future__ import annotations

import asyncio
import inspect
import logging
from dataclasses import dataclass
from itertools import count
from typing import Any, Callable

logger = logging.getLogger(__name__)

MAX_CONCURRENT_SUBSCRIBERS = 256

TopicUserConnected = "user:connected"
TopicUserDisconnected = "user:disconnected"
TopicMessageSent = "message:sent"
TopicMessageRead = "message:read"


@dataclass(slots=True)
class Event:
    topic: str
    data: Any


@dataclass(slots=True)
class _SubscriberEntry:
    id: int
    fn: Callable[[Event], Any]


class _Bus:
    def __init__(self) -> None:
        self._subscribers: dict[str, list[_SubscriberEntry]] = {}
        self._ids = count(1)
        self._semaphore = asyncio.Semaphore(MAX_CONCURRENT_SUBSCRIBERS)

    async def publish(self, topic: str, data: Any = None) -> None:
        subscribers = list(self._subscribers.get(topic, ()))
        if not subscribers:
            return

        event = Event(topic=topic, data=data)
        await asyncio.gather(*(self._run(entry.fn, event) for entry in subscribers), return_exceptions=True)

    def subscribe(self, topic: str, sub: Callable[[Event], Any]) -> Callable[[], None]:
        entry = _SubscriberEntry(next(self._ids), sub)
        self._subscribers.setdefault(topic, []).append(entry)

        def unsubscribe() -> None:
            subscribers = self._subscribers.get(topic, [])
            self._subscribers[topic] = [item for item in subscribers if item.id != entry.id]

        return unsubscribe

    async def _run(self, fn: Callable[[Event], Any], event: Event) -> None:
        async with self._semaphore:
            try:
                result = fn(event)
                if inspect.isawaitable(result):
                    await result
            except Exception:
                logger.exception("event subscriber failed: topic=%s subscriber=%s", event.topic, getattr(fn, "__name__", repr(fn)))

    def snapshot(self, topic: str | None = None) -> dict[str, list[Callable[[Event], Any]]]:
        if topic is not None:
            return {topic: [entry.fn for entry in self._subscribers.get(topic, [])]}
        return {name: [entry.fn for entry in entries] for name, entries in self._subscribers.items()}


DefaultBus = _Bus()


def subscribe(topic: str, sub: Callable[[Event], Any] | None = None):
    if sub is not None:
        return DefaultBus.subscribe(topic, sub)

    def decorator(fn: Callable[[Event], Any]) -> Callable[[Event], Any]:
        DefaultBus.subscribe(topic, fn)
        return fn

    return decorator


async def publish(topic: str, data: Any = None, **kwargs: Any) -> None:
    payload = data if data is not None else kwargs
    await DefaultBus.publish(topic, payload)


class EventBus:
    @staticmethod
    def publish(topic: str, data: Any = None) -> Any:
        return DefaultBus.publish(topic, data)

    @staticmethod
    def subscribe(topic: str, sub: Callable[[Event], Any]) -> Callable[[], None]:
        return DefaultBus.subscribe(topic, sub)

    @staticmethod
    def snapshot(topic: str | None = None) -> dict[str, list[Callable[[Event], Any]]]:
        return DefaultBus.snapshot(topic)
